test-data/unit/check-redefine.test

```diff
@@ -0,0 +1,451 @@
+tmp/m.py:4: error: Revealed type is 'builtins.object'
+tmp/m.py:6: error: Revealed type is 'builtins.str'
+main:3: error: Revealed type is 'builtins.str'
+
+[case testRedefineGlobalForIndex]
+# flags: --allow-redefinition
+import m
+reveal_type(m.x)
+[file m.py]
+from typing import Iterable
+def f(): pass
+it1: Iterable[int] = f()
+it2: Iterable[str] = f()
+for x in it1:
+    reveal_type(x)
+for x in it2:
+    reveal_type(x)
+reveal_type(x)
+[out]
+tmp/m.py:6: error: Revealed type is 'builtins.int*'
+tmp/m.py:8: error: Revealed type is 'builtins.str*'
+tmp/m.py:9: error: Revealed type is 'builtins.str*'
+main:3: error: Revealed type is 'builtins.str*'
+
+[case testRedefineGlobalBasedOnPreviousValues]
+# flags: --allow-redefinition
+from typing import TypeVar, Iterable
+T = TypeVar('T')
+def f(x: T) -> Iterable[T]: pass
+a = 0
+a = f(a)
+reveal_type(a) # E: Revealed type is 'typing.Iterable[builtins.int*]'
+
+[case testRedefineGlobalWithSeparateDeclaration]
+# flags: --allow-redefinition
+x = ''
+reveal_type(x) # E: Revealed type is 'builtins.str'
+x: int
+x = '' # E: Incompatible types in assignment (expression has type "str", variable has type "int")
+reveal_type(x) # E: Revealed type is 'builtins.int'
+x: object
+x = 1
+reveal_type(x) # E: Revealed type is 'builtins.int'
+if int():
+    x = object()
+
+[case testRedefineGlobalUsingForLoop]
+# flags: --allow-redefinition
+from typing import Iterable, TypeVar, Union
+T = TypeVar('T')
+def f(x: T) -> Iterable[Union[T, str]]: pass
+x = 0
+reveal_type(x)  # E: Revealed type is 'builtins.int'
+for x in f(x):
+    pass
+reveal_type(x)  # E: Revealed type is 'Union[builtins.int*, builtins.str]'
+
+[case testNoRedefinitionIfOnlyInitialized]
+# flags: --allow-redefinition --no-strict-optional
+x = None  # type: int
+x = '' # E: Incompatible types in assignment (expression has type "str", variable has type "int")
+x = object() # E: Incompatible types in assignment (expression has type "object", variable has type "int")
+x  # Reference to variable
+x = ''
+
+y = 0
+y = '' # E: Incompatible types in assignment (expression has type "str", variable has type "int")
+
+[case testNoRedefinitionIfNoValueAssigned]
+# flags: --allow-redefinition
+x: int
+x = '' # E: Incompatible types in assignment (expression has type "str", variable has type "int")
+reveal_type(x) # E: Revealed type is 'builtins.int'
+x: object
+
+[case testNoRedefinitionIfExplicitlyDisallowed]
+# flags: --disallow-redefinition
+x = 0
+x = 2
+x = ''  # E: Incompatible types in assignment (expression has type "str", variable has type "int")
+def f() -> None:
+    y = 0
+    y = '' # E: Incompatible types in assignment (expression has type "str", variable has type "int")
+class C:
+    y = 0
+    y = '' # E: Incompatible types in assignment (expression has type "str", variable has type "int")
+def g() -> None:
+    # _ is a special case
+    _ = 0
+    _ = ''
+    x, _ = 0, C()
+[builtins fixtures/tuple.pyi]
```

Comment:
I would add test cases for
```python
e = 1
e
try:
    ...
except Exception as e:
    ...
```
and
```python
with open('file') as f:  # IO[str]
    ...
with open('file', 'b') as f:  # IO[bytes]
    ...
```
(maybe plus some variations).

Reply:
Added test cases.